packages/core-p2p/src/peer-verifier.ts

```diff
@@ -0,0 +1,512 @@
+     *
+     * Case1. Peer height > our height and our highest block is part of the peer's chain.
+     *   This means the peer is ahead of us, on the same chain. No fork.
+     *   We verify: his blocks that have height > our height (up to the round end).
+     *
+     * Case2. Peer height > our height and our highest block is not part of the peer's chain.
+     *   This means that the peer is on a different, higher chain. It has forked before our
+     *   latest block.
+     *   We verify: the first few of the peer's blocks after the fork (up to the round end).
+     *
+     * Case3. Peer height == our height and our latest blocks are the same.
+     *   This means that our chains are the same.
+     *   We verify: nothing.
+     *
+     * Case4. Peer height == our height and our latest blocks differ.
+     *   This means that we are on a different chains with equal height. A fork has occurred.
+     *   We verify: same as 2.
+     *
+     * Case5. Peer height < our height and peer's latest block is part of our chain.
+     *   This means that the peer is on the same chain as us, just lagging behind.
+     *   We verify: nothing.
+     *
+     * Case6. Peer height < our height and peer's latest block is not part of our chain.
+     *   This means that we have forked and the peer's chain is lower.
+     *   We verify: same as 2.
+     *
+     * @param {Object} claimedState the claimed state of the peer, as returned by `/peer/status`
+     * @param {Number} deadline operation deadline, in milliseconds since Epoch
+     * @return {Boolean} true if the peer's blockchain is verified to be legit (albeit it may be
+     * different than our blockchain)
+     * @throws {Error} if the state verification could not complete before the deadline
+     */
+    public async checkState(claimedState: any, deadline: number): Promise<boolean> {
+        if (this.isStateInvalid(claimedState)) {
+            return false;
+        }
+
+        const ourHeight: number = await this.ourHeight();
+
+        if (await this.weHavePeersHighestBlock(claimedState, ourHeight)) {
+            // Case3 and Case5
+            return true;
+        }
+
+        const claimedHeight = Number(claimedState.header.height);
+
+        const highestCommonBlockHeight =
+            await this.findHighestCommonBlockHeight(claimedHeight, ourHeight, deadline);
+        if (highestCommonBlockHeight === null) {
+            return false;
+        }
+
+        if (!await this.verifyPeerBlocks(highestCommonBlockHeight + 1, claimedHeight, deadline)) {
+            return false;
+        }
+
+        this.logger.debug(`${this.logPrefix} success`);
+
+        return true;
+    }
+
+    /**
+     * Check if the state claimed by the peer is definitely invalid.
+     * @param {Object} claimedState peer's claimed state (from `/peer/status`)
+     * @return {Boolean} true if invalid
+     */
+    private isStateInvalid(claimedState: any): boolean {
```

Comment:
This looks wrong on so many levels.

The proper way would be to have schema validation for the
endpoint `/peer/status` and introduce an interface for the payload.

But it is out of scope of this PR, so just leave it like that for now.

Reply:
Right, a schema validation would be definitely superior to manually checking if some fields are present and have the expected types. Not sure about the interface for the payload. I will look into this after doing the tests for this or a followup PR.

Reply:
Actually the problem of manually checking fields instead of using schema validation is present in the code even before this patch. The code is peppered with such checks, for example `Peer.hasCommonBlocks()` has `body && body.success && body.common`. `Peer.getPeers()` assumes that the reply will contain `peers` and that it will be an array. `Peer.ping()` lacks any checks on what is returned.

I think the proper way to handle this is to extend `Peer.__get(endpoint, timeout)` with a 3rd argument - a schema that the reply must match. That would be out of scope for this PR, more suitable for a followup PR.

Reply:
Yeah, that's why I said it's out of scope. This is a general issue.

Reply:
Will do schema validation here and in all places (in peer.ts or even all of core-p2p) in a followup patch.